### PR TITLE
looking at ways to reduce validator registration calls

### PR DESCRIPTION
**Pull request by james-prysm**
- [x] I have read [CONTRIBUTING.md](CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

**What type of PR is this?**

other

**What does this PR do? Why is it needed?**

We should schedule validator registrations in the following conditions
- [x] when validator starts 
- [x] start of epoch ( send all validator registrations)
- [x] new account added 
- [x] existing account status change
- [x] proposer settings changed
- [x] beacon node reconnects

this pr reduces the amount of calls to check validator status as well as pushing validator registration. pushing proposer settings remains the same.

## Changes

- changes internal pubkey to validator index cache to pubkey to validator status: this will reduce the calls to check validator status as the status only updates at epoch boundaries
- adds updates to the pubkey to validator status cache on key reload
- only pushes validator registration at the start of the epoch or if the beacon node reconnects
    - only new keys will be sent while in between epochs

**Which issues(s) does this PR fix?**

Related to https://github.com/prysmaticlabs/prysm/issues/14369

**Other notes for review**

### CODE REVIEW

**Comment by james-prysm** [comment]
needs some tests

File: validator/client/validator.go [hunk 1]
```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
...
-				continue
-			}
-			v.pubkeyToValidatorIndex[k] = i
+			validatorStatusUnpopulated = true
+			// skip adding if not found
+			continue
 		}
 		copiedk := k
 		statusRequestKeys = append(statusRequestKeys, copiedk[:])
+		copiedStatus := sta
```
Comment (potuz) [inline validator/client/validator.go:1226]: Does this copy anything? isn´t  `sta` just a pointer?

File: validator/client/validator.go [hunk 2]
```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
...
-			}
-			if !ok { // Nothing we can do if RPC server doesn't have validator index.
-				continue
-			}
-			v.pubkeyToValidatorIndex[k] = i
+			validatorStatusUnpopulated = true
+			// skip adding if not found
+			continue
 		}
 		copiedk := k
```
Comment (potuz) [inline validator/client/validator.go:1224]: There's no need to copy on Go 1.22, proof https://go.dev/play/p/jfmrjZ2yfB2

File: validator/client/validator.go [hunk 3]
```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
...
+			sta := &validatorStatus{
+				publicKey: resp.PublicKeys[i],
+				status:    s,
+				index:     resp.Indices[i],
+			}
+			statuses = append(statuses, sta)
+			// update cache
+			v.pubkeyToStatus[bytesutil.ToBytes48(resp.PublicKeys[i])] = sta
+		}
+		validatorStatuses = statuses
```
Comment (potuz) [inline validator/client/validator.go:1256]: You `continue` in the previous loop, but anyway this will be overwritten here unconditionally

File: validator/client/validator.go [hunk 4]
```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
...
+		validatorStatuses = append(validatorStatuses, copiedStatus)
 	}
-	resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
-		PublicKeys: statusRequestKeys,
-	})
-	if err != nil {
-		return nil, err
+	if isEpochStart || validatorStatusUnpopulated {
+		resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
+			PublicKeys: statusRequestKeys,
```
Comment (potuz) [inline validator/client/validator.go:1231]: If the key didn't have a cached status, then it won't be in the `statusRequestKeys`, even though that key may have triggered this code path.

File: validator/client/validator.go [hunk 5]
```diff
@@ -1260,8 +1279,13 @@ func (v *validator) filterAndCacheActiveKeys(ctx context.Context, pubkeys [][fie
...
 	var prepareProposerReqs []*ethpb.PrepareBeaconProposerRequest_FeeRecipientContainer
 	for _, k := range activePubkeys {
+		s, ok := v.pubkeyToStatus[k]
+		if !ok {
+			continue
+		}
+
 		// Default case: Define fee recipient to burn address
-		var feeRecipient common.Address
+		feeRecipient := common.HexToAddress(params.BeaconConfig().EthBurnAddressHex)
```
Comment (potuz) [inline validator/client/validator.go:1288]: Why this change?

**Review by potuz — CHANGES_REQUESTED** [review]

File: validator/client/validator.go (hunk 5 above)
Comment (james-prysm) [reply validator/client/validator.go:1288]: this is configurable for the burn address at least by network, where as the other one should just always just go to the 0 address I think

File: validator/client/validator.go (hunk 1 above)
Comment (james-prysm) [reply validator/client/validator.go:1226]: I should just remove this

File: validator/client/validator.go (hunk 4 above)
Comment (james-prysm) [reply validator/client/validator.go:1231]: shoot yes, on epoch start i should skip the continue and the requests for all keys should be done anyways

File: validator/client/validator.go (hunk 3 above)
Comment (james-prysm) [reply validator/client/validator.go:1256]: This is only if it's on epoch start/only the validator statuses that changed in the middle of the epoch

File: validator/client/validator.go (hunk 4 above)
Comment (james-prysm) [reply validator/client/validator.go:1231]: been a few days I'm going to double check but I think the cache is populated already but been a few days so going to refresh myself and clean this up anyways

Comment (james-prysm) [reply validator/client/validator.go:1231]: definitely doing something wrong here will fix

Comment (james-prysm) [reply validator/client/validator.go:1228]: should not have had the continue on it...

File: validator/client/validator.go (hunk 3 above)
Comment (james-prysm) [reply validator/client/validator.go:1254]: having the continue was the bug here

**Review by potuz — CHANGES_REQUESTED** [review]
I think the logic should be different for `filterAndCacheActiveKeys`

1. if it's an epoch start, call the RPC and update statuses and return. 
2. If it's not an epoch start, get the statuses from the local cache, if there's one missing, call RPC and return. 

More importantly, I don't think you are actually using at all the new cache `pubkeyToStatus` in any meaningful way

**Review by potuz — APPROVED** [review]

File: validator/client/validator.go [hunk 6]
```diff
@@ -1212,56 +1209,81 @@ func (v *validator) ChangeHost() {
...
+	for _, k := range pubkeys {
+		statusRequestKeys = append(statusRequestKeys, k[:])
+	}
+	resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
+		PublicKeys: statusRequestKeys,
+	})
+	if err != nil {
+		return err
+	}
+	if resp == nil {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/validator.go:1256]: actually since it's an api call I think it can be

**Review by rkapka — APPROVED** [review]
